### PR TITLE
feat(esbuild-bundle-analyzer): Simple analyzer tool for esbuild bundles based on esbuild's metafile 

### PR DESCRIPTION
**Pull request by AAHAbbas**
### Description
This tool provides simple analysis of bundles based on [esbuild's metafile](https://esbuild.github.io/api/#metafile). It allows users to analyze a bundle/metafile, compare two metafiles, and generate a limited, but compatible Webpack stats file. This file can then be used with more advanced Webpack-based analysis and comparison tools such as [@mixer/webpack-bundle-compare](https://github.com/microsoft/webpack-bundle-compare). This tool could allow developers quickly and easily gain insights into their bundle size and composition.

Experimental!

### Test plan

This tool has three commands, `analyze`, `compare`, and `transform`.

#### analyze
`node lib/cli.js analyze meta.json --show-duplicates`
Analyzes the metafile `meta.json` and if the bundle contains duplicates, then it will output an in-depth detail about how and why the duplicates are being included in the bundle. 

#### compare
`node lib/cli.js compare --baseline meta1.json --candidate meta2.json`
Compares two metafiles and outputs the difference between the two.

#### transform
`node lib/cli.js transform meta.json --output webpack_stats.json`
Generates a Webpack stats file based on the metafile.

### CODE REVIEW

File: incubator/metro-analyzer/src/output.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    outputToConsole(result);
+  }
+}
+
+/**
+ * Outputs a list of duplicate dependencies to the console.
+ *
+ * @param duplicates  List of duplicate dependencies in the bundle
+ */
+export function outputDuplicates(duplicates: Duplicate[]): void {
```
Comment (tido64) [inline incubator/metro-analyzer/src/output.ts:126]: I'm wondering how feasible it is to reuse `metro-plugin-duplicates-checker` as much as possible. For instance, there's a function for outputting duplicates that you probably could reuse given the correct format.

**Review by tido64 — COMMENTED** [review]
Thanks for submitting this. This is pretty cool, enabling Webpack tooling for esbuild users. In general, I think we should keep this generic and not mention Metro as much. I think that would do it a disservice.

File: incubator/metro-analyzer/README.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
+<!-- We recommend an empty change log entry for a new package: `yarn change --empty` -->
+
+# @rnx-kit/metro-analyzer
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/README.md:3]: Yeah, I agree. When I first started working on this I only had Metro on mind. Do you have any suggestions? `esbuild-analyzer`?

File: incubator/metro-analyzer/src/cli.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            describe:
+              "Get detailed information about how the duplicates are bundled",
+            type: "boolean",
+            default: false,
+          })
+          .option("transform", {
+            describe:
+              "Generates a webpack stats file from the esbuild metafile and sets the output file to write the stats file to",
+            type: "string",
+          })
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/src/cli.ts:30]: Right, I only added it so I can chain analyzing and transforming the metafile in a single execution, but I can remove it.

File: incubator/metro-analyzer/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (tido64) [reply incubator/metro-analyzer/README.md:3]: `esbuild-bundle-analyzer`?

File: incubator/metro-analyzer/README.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧🚧
+
+This is a sample folder to use as base for generating new packages for
+`rnx-kit`.
+
+## Motivation
+
+We want new packages to follow an existing set of patterns and guidelines; via
+this package, we can enforce easily allow new folders to stick to at least a
+common starting point.
```
[the remark this answers is not in the document]
Comment (tido64) [reply incubator/metro-analyzer/README.md:21]: Can you also mention how this tool compares to https://esbuild.github.io/analyze/?

File: incubator/metro-analyzer/test/index.test.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+// TODO: add actual tests!
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/test/index.test.ts]: Added some tests 😄

File: incubator/metro-analyzer/src/output.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+          type = "Imported stylesheet file";
+          break;
+        case "url-token":
+          type = "URL reference file";
+          break;
+        default:
+          throw new Error(`Invalid import kind: ${item?.import?.kind}`);
+      }
+
+      console.log(
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/src/output.ts]: Switched to `info`.

File: incubator/metro-analyzer/src/cli.ts [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            describe: "The esbuild metafile to analyze",
+            type: "string",
+            demandOption: true,
+          })
+          .option("detailed", {
+            describe:
+              "Get detailed information about how the duplicates are bundled",
+            type: "boolean",
+            default: false,
+          })
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/src/cli.ts]: Renamed to `showDuplicates`. Fine?

File: incubator/metro-analyzer/src/cli.ts [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import { compare } from "./compare.js";
+import { webpackStats } from "./webpackStats.js";
+import type { Metafile } from "./metafile.js";
+import fs from "fs";
+
+export function main(): void {
+  yargs(process.argv.slice(2))
+    .command(
+      "analyze",
+      "Analyzes a metro bundle by analyzing the esbuild metafile",
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/src/cli.ts]: Removed all mentions of Metro

File: incubator/metro-analyzer/src/duplicates.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
+import { countCopies } from "@rnx-kit/metro-plugin-duplicates-checker/lib/checkForDuplicatePackages.js";
+import { resolveModule } from "@rnx-kit/metro-plugin-duplicates-checker/lib/gatherModules.js";
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/src/duplicates.ts]: Exported them.

File: incubator/metro-analyzer/src/analyze.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+ * @param jsonFile Output file to write analysis information to in JSON format
+ */
+export async function analyze(
+  metafilePath: string,
+  detailed: boolean,
+  transformPath?: string,
+  jsonFile?: string
+) {
+  const content = fs.readFileSync(metafilePath, { encoding: "utf-8" });
+  const metafile: Metafile = JSON.parse(content);
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/src/analyze.ts]: Replaced with `readMetafile` call

File: incubator/metro-analyzer/src/compare.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * @param baselinePath Path to the baseline metafile
+ * @param candidatePath Path to the candidate metafile
+ */
+export async function compare(
+  baselinePath: string,
+  candidatePath: string
+): Promise<void> {
+  const baseline = await readMetafile(baselinePath);
+  const candidate = await readMetafile(candidatePath);
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/metro-analyzer/src/compare.ts]: Changed to use `fs.readFileSync`.

File: incubator/esbuild-bundle-analyzer/src/analyze.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  if (errorMessage) {
+    error(errorMessage);
+  }
+
+  if (errorMessage && showDuplicates) {
+    outputWhyDuplicateInBundle(
+      getWhyDuplicatesInBundle(metafile, graph),
+      namespace
+    );
+  }
```
Comment (tido64) [inline incubator/esbuild-bundle-analyzer/src/analyze.ts:53]:
```suggestion
  if (errorMessage) {
    error(errorMessage);
    if (showDuplicates) {
      outputWhyDuplicateInBundle(
        getWhyDuplicatesInBundle(metafile, graph),
        namespace
      );
    }
  }
```

File: incubator/esbuild-bundle-analyzer/src/cli.ts [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+            describe: "The esbuild metafile to analyze",
+            type: "string",
+            demandOption: true,
+          })
+          .option("outputFile", {
+            describe: "Output file to write the Webpack stats file to",
+            type: "string",
+            demandOption: true,
+          })
+          .option("skipLineNumber", {
```
Comment (tido64) [inline incubator/esbuild-bundle-analyzer/src/cli.ts:82]:
```suggestion
          .option("skip-line-number", {
```

File: incubator/esbuild-bundle-analyzer/src/analyze.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+  const data = stats(metafile);
+  output(
+    {
+      data: data,
+      buildTime: 0,
+      downloadTime: Math.round((data.totalBytes * 8) / DOWNLOAD_SPEED),
```
Comment (tido64) [inline incubator/esbuild-bundle-analyzer/src/analyze.ts:60]: Suggestion: We should either make this configurable or show multiple numbers like Bundlephobia does:
Screenshot 2023-06-20 at 15 40 15

File: incubator/esbuild-bundle-analyzer/src/webpackStats.ts [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    });
+
+    return lineNumber !== -1 ? (lineNumber + 1).toString() : "";
+  } catch (err) {
+    return "";
+  }
+}
+
+/** Returns a simpler and more readable path which starts from node_modules, e.g.:
+ * ../../.store/@test-library@0.0.1-b55dbe3d1aed7a6c074d/node_modules/@test-library/a/b/test.js
```
Comment (tido64) [inline incubator/esbuild-bundle-analyzer/src/webpackStats.ts:43]: Nit:
```suggestion
/**
 * Returns a simpler and more readable path which starts from node_modules, e.g.:
 * ../../.store/@test-library@0.0.1-b55dbe3d1aed7a6c074d/node_modules/@test-library/a/b/test.js
```

File: incubator/esbuild-bundle-analyzer/src/webpackStats.ts [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        return line
+          .replace("/lib/", "/src/")
+          .replace(".js", ".ts")
+          .includes(keyword);
+      }
+
+      return result;
+    });
+
+    return lineNumber !== -1 ? (lineNumber + 1).toString() : "";
```
Comment (tido64) [inline incubator/esbuild-bundle-analyzer/src/webpackStats.ts:36]: In general, it's better compare against 0:
```suggestion
    return lineNumber >= 0 ? (lineNumber + 1).toString() : "";
```

File: incubator/esbuild-bundle-analyzer/src/cli.ts [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+            describe:
+              "Get detailed information about how the duplicates are bundled",
+            type: "boolean",
+            default: false,
+          })
+          .option("transform", {
+            describe:
+              "Generates a webpack stats file from the esbuild metafile and sets the output file to write the stats file to",
+            type: "string",
+          })
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/esbuild-bundle-analyzer/src/cli.ts]: Removed.

File: incubator/esbuild-bundle-analyzer/src/stats.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+  for (const output in outputs) {
+    // Don't count sourcemaps
+    if (!output.includes(".map")) {
+      countOut++;
+      bytesOut += outputs[output].bytes;
+    }
+  }
+
+  return {
+    files: Object.keys(inputs).length,
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/esbuild-bundle-analyzer/src/stats.ts]: Done

File: incubator/esbuild-bundle-analyzer/src/analyze.ts (hunk 14 above)
Comment (AAHAbbas) [reply incubator/esbuild-bundle-analyzer/src/analyze.ts]: Updated to show multiple numbers like Bundlephobia (slow 3G and fast 4G). We can probably make this more configurable in the future.

File: incubator/esbuild-bundle-analyzer/README.md [hunk 19]
```diff
@@ -55,18 +55,23 @@ npx @rnx-kit/esbuild-bundle-analyzer analyze --metafile <path-to-esbuild-metafil
...
+- `--show-duplicates`, Boolean flag to output how each duplicated file is being
+  added to the bundle
 - `--json`, Outputs the analysis in JSON format and sets the output file to
   write the analysis information to. If not set the analysis will be output to
   the console
+- `--namespace`, Every module has an associated namespace. By default esbuild
+  operates in the file namespace, which corresponds to files on the `file`
+  system.
+  [@rnx-kit/metro-serializer-esbuild](https://github.com/microsoft/rnx-kit/tree/main/packages/metro-serializer-esbuild)
+  uses the `virtual:metro` namespace. This flag allows you to specify the
```
[the remark this answers is not in the document]
Comment (AAHAbbas) [reply incubator/esbuild-bundle-analyzer/README.md:67]: Aha I see, I'll note this issue down with the rest of the issues/todos.

File: .changeset/breezy-mayflies-vanish.md [hunk 20]
```diff
@@ -0,0 +1,7 @@
+---
+"@rnx-kit/esbuild-bundle-analyzer": minor
+"@rnx-kit/metro-plugin-cyclic-dependencies-detector": patch
+"@rnx-kit/metro-plugin-duplicates-checker": patch
```
Comment (tido64) [inline .changeset/breezy-mayflies-vanish.md:4]: Can you move these to separate change files and add the appropriate change log entry?

File: .changeset/sharp-crabs-tap.md [hunk 21]
```diff
@@ -0,0 +1,6 @@
+---
+"@rnx-kit/metro-plugin-cyclic-dependencies-detector": patch
+"@rnx-kit/metro-plugin-duplicates-checker": patch
+---
+
+Export functions
```
Comment (tido64) [inline .changeset/sharp-crabs-tap.md:6]: Can we split this into separate files and add more details 😄

File: incubator/esbuild-bundle-analyzer/package.json [hunk 22]
```diff
@@ -0,0 +1,61 @@
+{
+  "private": true,
```
Comment (tido64) [inline incubator/esbuild-bundle-analyzer/package.json:2]: Btw, if you want to start publishing this package, you should remove this prop.

Comment (AAHAbbas) [reply incubator/esbuild-bundle-analyzer/package.json]: Removed!

**Review by tido64 — APPROVED** [review]
giphy

**Pull request merged**